In [ ]:
import os
path="/content/drive/MyDrive/cointab_assignment/data"
os.listdir(path)



['Company X - Pincode Zones.xlsx',
 'Courier Company - Invoice.xlsx',
 'Courier Company - Rates.xlsx',
 'Company X - Order Report.xlsx',
 'Company X - SKU Master.xlsx',
 'Expected_Result.xlsx']

In [ ]:
import pandas as pd 


### Company X - Order Report.xlsx

In [ ]:
comp_ord=pd.read_excel("/content/drive/MyDrive/cointab_assignment/data/Company X - Order Report.xlsx")

In [ ]:
comp_ord

ExternOrderNo            SKU  Order Qty
0       2001827036  8904223818706        1.0
1       2001827036  8904223819093        1.0
2       2001827036  8904223819109        1.0
3       2001827036  8904223818430        1.0
4       2001827036  8904223819277        1.0
..             ...            ...        ...
395     2001806229  8904223818942        1.0
396     2001806229  8904223818850        1.0
397     2001806226  8904223818850        2.0
398     2001806210  8904223816214        1.0
399     2001806210  8904223818874        1.0

[400 rows x 3 columns]

In [ ]:
comp_ord.isnull().sum()

ExternOrderNo    0
SKU              0
Order Qty        0
dtype: int64

In [ ]:
comp_ord.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ExternOrderNo  400 non-null    int64  
 1   SKU            400 non-null    object 
 2   Order Qty      400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


In [ ]:
comp_ord.describe()

ExternOrderNo   Order Qty
count   4.000000e+02  400.000000
mean    2.001811e+09    1.297500
std     5.187698e+03    0.919842
min     2.001806e+09    1.000000
25%     2.001807e+09    1.000000
50%     2.001809e+09    1.000000
75%     2.001812e+09    1.000000
max     2.001827e+09    8.000000

In [ ]:
[print(comp_ord[i].nunique()) for i in comp_ord.columns]

124
65
6


[None, None, None]

#Analysis 

1. There are toal of 400 rows of which we can get the following 
  - There are total of 124 orders 
  - There are total of 65 products
  - The maximum weight of a single product is 8 Nos

### Company X - SKU Master.xlsx

In [ ]:
comp_sku=pd.read_excel("/content/drive/MyDrive/cointab_assignment/data/Company X - SKU Master.xlsx")

In [ ]:
comp_sku

SKU  Weight (g)
0   8904223815682         210
1   8904223815859         165
2   8904223815866         113
3   8904223815873          65
4   8904223816214         120
..            ...         ...
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

[66 rows x 2 columns]

In [ ]:
comp_sku.shape

(66, 2)

In [ ]:
comp_sku.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   SKU         66 non-null     object
 1   Weight (g)  66 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.2+ KB


In [ ]:
comp_sku.describe()

Weight (g)
count   66.000000
mean   219.742424
std    146.855482
min     10.000000
25%    116.250000
50%    165.000000
75%    290.000000
max    600.000000

In [ ]:
comp_sku.isnull().sum()

SKU           0
Weight (g)    0
dtype: int64

In [ ]:
[print(comp_sku[i].nunique()) for i in comp_sku.columns]

65
31


[None, None]

In [ ]:
comp_sku.drop_duplicates(inplace=True)

In [ ]:
comp_sku


SKU  Weight (g)
0   8904223815682         210
1   8904223815859         165
2   8904223815866         113
3   8904223815873          65
4   8904223816214         120
..            ...         ...
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

[65 rows x 2 columns]

#Analysis 

The total unique products matched with the no of items with Orders Table 

### Company X - Pincode Zones.xlsx

In [ ]:
comp_pin_zone=pd.read_excel("/content/drive/MyDrive/cointab_assignment/data/Company X - Pincode Zones.xlsx")

In [ ]:
comp_pin_zone

Warehouse Pincode  Customer Pincode Zone
0               121003            507101    d
1               121003            486886    d
2               121003            532484    d
3               121003            143001    b
4               121003            515591    d
..                 ...               ...  ...
119             121003            325207    b
120             121003            303702    b
121             121003            313301    b
122             121003            173212    e
123             121003            302020    b

[124 rows x 3 columns]

In [ ]:
comp_pin_zone.isnull().sum()

Warehouse Pincode    0
Customer Pincode     0
Zone                 0
dtype: int64

In [ ]:
comp_pin_zone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Warehouse Pincode  124 non-null    int64 
 1   Customer Pincode   124 non-null    int64 
 2   Zone               124 non-null    object
dtypes: int64(2), object(1)
memory usage: 3.0+ KB


In [ ]:
[print(comp_pin_zone[i].nunique()) for i in comp_pin_zone.columns]

1
108
3


[None, None, None]

In [ ]:
comp_pin_zone.drop_duplicates(inplace=True)

In [ ]:
comp_pin_zone.shape

(108, 3)

#Analysis 

There are a total of 108 unique pincodes to be serviced by the courier company 

In [ ]:
cour_invoc=pd.read_excel("/content/drive/MyDrive/cointab_assignment/data/Courier Company - Invoice.xlsx")

In [ ]:
cour_invoc

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091118551656  2001812941            0.73             121003   
120  1091117614452  2001809383            0.50             121003   
121  1091120922803  2001820978            0.50             121003   
122  1091121844806  2001811475            0.50             121003   
123  1091121846136  2001811305            0.50             121003   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
0              507101    d          Forward charges                 135.0  
1              486886    d          Forward charges                  90.2  
2              532484    d          Forward charges                 224.6  
3              143001    b          Forward charges                  61.3  
4              515591    d          Forward charges                  45.4  
..                ...  ...                      ...                   ...  
119            325207    d          Forward charges                  90.2  
120            303702    d  Forward and RTO charges                  86.7  
121            313301    d          Forward charges                  45.4  
122            173212    b          Forward charges                  33.0  
123            302020    d          Forward charges                  45.4  

[124 rows x 8 columns]

In [ ]:
cour_invoc.isnull().sum()

AWB Code                0
Order ID                0
Charged Weight          0
Warehouse Pincode       0
Customer Pincode        0
Zone                    0
Type of Shipment        0
Billing Amount (Rs.)    0
dtype: int64

In [ ]:
cour_invoc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AWB Code              124 non-null    int64  
 1   Order ID              124 non-null    int64  
 2   Charged Weight        124 non-null    float64
 3   Warehouse Pincode     124 non-null    int64  
 4   Customer Pincode      124 non-null    int64  
 5   Zone                  124 non-null    object 
 6   Type of Shipment      124 non-null    object 
 7   Billing Amount (Rs.)  124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


In [ ]:
cour_invoc.describe()

AWB Code      Order ID  Charged Weight  Warehouse Pincode  \
count  1.240000e+02  1.240000e+02      124.000000              124.0   
mean   1.091118e+12  2.001811e+09        0.956048           121003.0   
std    1.473661e+06  5.167329e+03        0.662815                0.0   
min    1.091117e+12  2.001806e+09        0.150000           121003.0   
25%    1.091117e+12  2.001807e+09        0.667500           121003.0   
50%    1.091117e+12  2.001809e+09        0.725000           121003.0   
75%    1.091119e+12  2.001812e+09        1.100000           121003.0   
max    1.091122e+12  2.001827e+09        4.130000           121003.0   

       Customer Pincode  Billing Amount (Rs.)  
count        124.000000            124.000000  
mean      365488.072581            110.066129  
std       152156.322130             64.060832  
min       140301.000000             33.000000  
25%       302017.000000             86.700000  
50%       321304.500000             90.200000  
75%       405102.250000            135.000000  
max       845438.000000            403.800000

In [ ]:
cour_invoc.shape

(124, 8)

In [ ]:
[print(i, ": ",cour_invoc[i].nunique()) for i in cour_invoc.columns]

AWB Code :  124
Order ID :  124
Charged Weight :  54
Warehouse Pincode :  1
Customer Pincode :  108
Zone :  3
Type of Shipment :  2
Billing Amount (Rs.) :  20


[None, None, None, None, None, None, None, None]

In [ ]:
cour_invoc

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091118551656  2001812941            0.73             121003   
120  1091117614452  2001809383            0.50             121003   
121  1091120922803  2001820978            0.50             121003   
122  1091121844806  2001811475            0.50             121003   
123  1091121846136  2001811305            0.50             121003   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
0              507101    d          Forward charges                 135.0  
1              486886    d          Forward charges                  90.2  
2              532484    d          Forward charges                 224.6  
3              143001    b          Forward charges                  61.3  
4              515591    d          Forward charges                  45.4  
..                ...  ...                      ...                   ...  
119            325207    d          Forward charges                  90.2  
120            303702    d  Forward and RTO charges                  86.7  
121            313301    d          Forward charges                  45.4  
122            173212    b          Forward charges                  33.0  
123            302020    d          Forward charges                  45.4  

[124 rows x 8 columns]

###  'Courier Company - Rates.xlsx',


In [ ]:
cour_rates=pd.read_excel("/content/drive/MyDrive/cointab_assignment/data/Courier Company - Rates.xlsx")

In [ ]:
cour_rates

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [ ]:
cour_rates=cour_rates.transpose()


In [ ]:
cour_rates.shape

(20, 1)

In [ ]:
# Creating Additional comumns

In [ ]:
cour_rates.rename(columns = {0:'rate'}, inplace = True)

In [ ]:
cour_rates

rate
fwd_a_fixed       29.5
fwd_a_additional  23.6
fwd_b_fixed       33.0
fwd_b_additional  28.3
fwd_c_fixed       40.1
fwd_c_additional  38.9
fwd_d_fixed       45.4
fwd_d_additional  44.8
fwd_e_fixed       56.6
fwd_e_additional  55.5
rto_a_fixed       13.6
rto_a_additional  23.6
rto_b_fixed       20.5
rto_b_additional  28.3
rto_c_fixed       31.9
rto_c_additional  38.9
rto_d_fixed       41.3
rto_d_additional  44.8
rto_e_fixed       50.7
rto_e_additional  55.5

In [ ]:
cour_rates["desc"]=cour_rates.index

In [ ]:
cour_rates

rate              desc
fwd_a_fixed       29.5       fwd_a_fixed
fwd_a_additional  23.6  fwd_a_additional
fwd_b_fixed       33.0       fwd_b_fixed
fwd_b_additional  28.3  fwd_b_additional
fwd_c_fixed       40.1       fwd_c_fixed
fwd_c_additional  38.9  fwd_c_additional
fwd_d_fixed       45.4       fwd_d_fixed
fwd_d_additional  44.8  fwd_d_additional
fwd_e_fixed       56.6       fwd_e_fixed
fwd_e_additional  55.5  fwd_e_additional
rto_a_fixed       13.6       rto_a_fixed
rto_a_additional  23.6  rto_a_additional
rto_b_fixed       20.5       rto_b_fixed
rto_b_additional  28.3  rto_b_additional
rto_c_fixed       31.9       rto_c_fixed
rto_c_additional  38.9  rto_c_additional
rto_d_fixed       41.3       rto_d_fixed
rto_d_additional  44.8  rto_d_additional
rto_e_fixed       50.7       rto_e_fixed
rto_e_additional  55.5  rto_e_additional

In [ ]:
cour_rates["type"]=cour_rates["desc"].apply(lambda x:x[:3])

In [ ]:
cour_rates["zone_code"]=cour_rates["desc"].apply(lambda x:x[4:5])

In [ ]:
cour_rates["rate_type"]=cour_rates["desc"].apply(lambda x:x[6:])

In [ ]:
cour_rates

rate              desc type zone_code   rate_type
fwd_a_fixed       29.5       fwd_a_fixed  fwd         a       fixed
fwd_a_additional  23.6  fwd_a_additional  fwd         a  additional
fwd_b_fixed       33.0       fwd_b_fixed  fwd         b       fixed
fwd_b_additional  28.3  fwd_b_additional  fwd         b  additional
fwd_c_fixed       40.1       fwd_c_fixed  fwd         c       fixed
fwd_c_additional  38.9  fwd_c_additional  fwd         c  additional
fwd_d_fixed       45.4       fwd_d_fixed  fwd         d       fixed
fwd_d_additional  44.8  fwd_d_additional  fwd         d  additional
fwd_e_fixed       56.6       fwd_e_fixed  fwd         e       fixed
fwd_e_additional  55.5  fwd_e_additional  fwd         e  additional
rto_a_fixed       13.6       rto_a_fixed  rto         a       fixed
rto_a_additional  23.6  rto_a_additional  rto         a  additional
rto_b_fixed       20.5       rto_b_fixed  rto         b       fixed
rto_b_additional  28.3  rto_b_additional  rto         b  additional
rto_c_fixed       31.9       rto_c_fixed  rto         c       fixed
rto_c_additional  38.9  rto_c_additional  rto         c  additional
rto_d_fixed       41.3       rto_d_fixed  rto         d       fixed
rto_d_additional  44.8  rto_d_additional  rto         d  additional
rto_e_fixed       50.7       rto_e_fixed  rto         e       fixed
rto_e_additional  55.5  rto_e_additional  rto         e  additional

## Resultant Table
    ● Order ID
    ● AWB Number
    ● Total weight as per X (KG)
    ● Weight slab as per X (KG)
    ● Total weight as per Courier Company (KG)
    ● Weight slab charged by Courier Company (KG)
    ● Delivery Zone as per X
    ● Delivery Zone charged by Courier Company
    ● Expected Charge as per X (Rs.)
    ● Charges Billed by Courier Company (Rs.)
    ● Difference Between Expected Charges and Billed Charges (Rs.)

In [ ]:
df=pd.DataFrame()

In [ ]:
df["Order ID"]=cour_invoc["Order ID"]
df["AWB Number"]=cour_invoc["AWB Code"]


In [ ]:
def total_comp_wt(orderid):
  temp_df= comp_ord[comp_ord["ExternOrderNo"]==orderid]
  temp_df=pd.merge(temp_df,comp_sku,on='SKU',how='left')
  temp_df["Total_wt"]=temp_df["Weight (g)"]*temp_df["Order Qty"]
  return ( temp_df["Total_wt"].sum())/1000


df["Total weight as per X KG"]=df["Order ID"].apply(total_comp_wt)
df

Order ID     AWB Number  Total weight as per X KG
0    2001806232  1091117222124                     1.302
1    2001806273  1091117222194                     0.615
2    2001806408  1091117222931                     2.265
3    2001806458  1091117223244                     0.700
4    2001807012  1091117229345                     0.240
..          ...            ...                       ...
119  2001812941  1091118551656                     0.500
120  2001809383  1091117614452                     0.607
121  2001820978  1091120922803                     0.515
122  2001811475  1091121844806                     0.689
123  2001811305  1091121846136                     0.750

[124 rows x 3 columns]

In [ ]:
def get_wt_slab(weight):
 weight=weight*1000
 ct=0
 while(weight >500):
   ct+=1
   
   weight=weight-500
 ct=ct+1
 return (ct*500)/1000

df["Weight slab as per X KG"]=df["Total weight as per X KG"].apply(get_wt_slab)
df


Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  
0                        1.5  
1                        1.0  
2                        2.5  
3                        1.0  
4                        0.5  
..                       ...  
119                      0.5  
120                      1.0  
121                      1.0  
122                      1.0  
123                      1.0  

[124 rows x 4 columns]

In [ ]:
def total_cour_wt(orderid):
  k=cour_invoc[cour_invoc["Order ID"] == orderid]["Charged Weight"].values
  return k[0]

df["Total weight as per Courier Company KG"]=df["Order ID"].apply(total_cour_wt)
df

Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  
0                        1.5                                    1.30  
1                        1.0                                    1.00  
2                        2.5                                    2.50  
3                        1.0                                    1.00  
4                        0.5                                    0.15  
..                       ...                                     ...  
119                      0.5                                    0.73  
120                      1.0                                    0.50  
121                      1.0                                    0.50  
122                      1.0                                    0.50  
123                      1.0                                    0.50  

[124 rows x 5 columns]

In [ ]:

df[" Weight slab charged by Courier Company KG"]=df["Total weight as per Courier Company KG"].apply(get_wt_slab)
df

Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  \
0                        1.5                                    1.30   
1                        1.0                                    1.00   
2                        2.5                                    2.50   
3                        1.0                                    1.00   
4                        0.5                                    0.15   
..                       ...                                     ...   
119                      0.5                                    0.73   
120                      1.0                                    0.50   
121                      1.0                                    0.50   
122                      1.0                                    0.50   
123                      1.0                                    0.50   

      Weight slab charged by Courier Company KG  
0                                           1.5  
1                                           1.0  
2                                           2.5  
3                                           1.0  
4                                           0.5  
..                                          ...  
119                                         1.0  
120                                         0.5  
121                                         0.5  
122                                         0.5  
123                                         0.5  

[124 rows x 6 columns]

In [ ]:
def company_zone(orderid):
  pincode=cour_invoc[cour_invoc["Order ID"]==orderid]["Customer Pincode"].iloc[0]
  return comp_pin_zone[comp_pin_zone["Customer Pincode"]==pincode ]["Comap_Zone"].iloc[0]

df["Company_Zone"]=df["Order ID"].apply(company_zone)
df


Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  \
0                        1.5                                    1.30   
1                        1.0                                    1.00   
2                        2.5                                    2.50   
3                        1.0                                    1.00   
4                        0.5                                    0.15   
..                       ...                                     ...   
119                      0.5                                    0.73   
120                      1.0                                    0.50   
121                      1.0                                    0.50   
122                      1.0                                    0.50   
123                      1.0                                    0.50   

      Weight slab charged by Courier Company KG Company_Zone  
0                                           1.5            d  
1                                           1.0            d  
2                                           2.5            d  
3                                           1.0            b  
4                                           0.5            d  
..                                          ...          ...  
119                                         1.0            b  
120                                         0.5            b  
121                                         0.5            b  
122                                         0.5            e  
123                                         0.5            b  

[124 rows x 7 columns]

In [ ]:
def courier_zone(orderid):
  return cour_invoc[cour_invoc["Order ID"]==orderid]["Zone"].iloc[0]

df["Courier_Zone"]=df["Order ID"].apply(company_zone)
df

Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  \
0                        1.5                                    1.30   
1                        1.0                                    1.00   
2                        2.5                                    2.50   
3                        1.0                                    1.00   
4                        0.5                                    0.15   
..                       ...                                     ...   
119                      0.5                                    0.73   
120                      1.0                                    0.50   
121                      1.0                                    0.50   
122                      1.0                                    0.50   
123                      1.0                                    0.50   

      Weight slab charged by Courier Company KG Company_Zone Courier_Zone  
0                                           1.5            d            d  
1                                           1.0            d            d  
2                                           2.5            d            d  
3                                           1.0            b            b  
4                                           0.5            d            d  
..                                          ...          ...          ...  
119                                         1.0            b            b  
120                                         0.5            b            b  
121                                         0.5            b            b  
122                                         0.5            e            e  
123                                         0.5            b            b  

[124 rows x 8 columns]

In [ ]:
def get_company_charges(orderid):
  order_type=cour_invoc[cour_invoc["Order ID"]==orderid]["Type of Shipment"].iloc[0]
  company_wt_slab=df[df["Order ID"]==orderid]["Weight slab as per X KG"].iloc[0]
  company_zone=df[df["Order ID"]==orderid]["Company_Zone"].iloc[0]
  weight_sec=company_wt_slab/0.5
  
  fwd_base_chrg=cour_rates[(cour_rates["type"]=="fwd" ) & (cour_rates["zone_code"]==company_zone) & (cour_rates["rate_type"]=="fixed")]["rate"].iloc[0]
  fwd_addl_chrg=cour_rates[(cour_rates["type"]=="fwd" ) & (cour_rates["zone_code"]==company_zone) & (cour_rates["rate_type"]=="additional")]["rate"].iloc[0]

  amount=0
  if order_type=="Forward charges":
    
    amount+=fwd_base_chrg
    amount+=(fwd_addl_chrg*(weight_sec-1))
    return amount
  elif order_type=="Forward and RTO charges":
    rto_base_chrg=cour_rates[(cour_rates["type"]=="rto" ) & (cour_rates["zone_code"]==company_zone) & (cour_rates["rate_type"]=="fixed")]["rate"].iloc[0]
    rto_addl_chrg=cour_rates[(cour_rates["type"]=="rto" ) & (cour_rates["zone_code"]==company_zone) & (cour_rates["rate_type"]=="additional")]["rate"].iloc[0]

    amount+=fwd_base_chrg
    amount+=(fwd_addl_chrg*(weight_sec-1))
    amount+=rto_base_chrg
    amount+=(rto_addl_chrg*(weight_sec-1))
    return amount

df["Expected Charge as per X (Rs.)"]=df["Order ID"].apply(get_company_charges)
df

Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  \
0                        1.5                                    1.30   
1                        1.0                                    1.00   
2                        2.5                                    2.50   
3                        1.0                                    1.00   
4                        0.5                                    0.15   
..                       ...                                     ...   
119                      0.5                                    0.73   
120                      1.0                                    0.50   
121                      1.0                                    0.50   
122                      1.0                                    0.50   
123                      1.0                                    0.50   

      Weight slab charged by Courier Company KG Company_Zone Courier_Zone  \
0                                           1.5            d            d   
1                                           1.0            d            d   
2                                           2.5            d            d   
3                                           1.0            b            b   
4                                           0.5            d            d   
..                                          ...          ...          ...   
119                                         1.0            b            b   
120                                         0.5            b            b   
121                                         0.5            b            b   
122                                         0.5            e            e   
123                                         0.5            b            b   

     Expected Charge as per X (Rs.)  
0                             135.0  
1                              90.2  
2                             224.6  
3                              61.3  
4                              45.4  
..                              ...  
119                            33.0  
120                           110.1  
121                            61.3  
122                           112.1  
123                            61.3  

[124 rows x 9 columns]

In [ ]:
def get_courier_charges(orderid):
  order_type=cour_invoc[cour_invoc["Order ID"]==orderid]["Type of Shipment"].iloc[0]
  Courier_wt_slab=df[df["Order ID"]==orderid][" Weight slab charged by Courier Company KG"].iloc[0]
  Courier_zone=df[df["Order ID"]==orderid]["Courier_Zone"].iloc[0]
  weight_sec=Courier_wt_slab/0.5
  
  fwd_base_chrg=cour_rates[(cour_rates["type"]=="fwd" ) & (cour_rates["zone_code"]==Courier_zone) & (cour_rates["rate_type"]=="fixed")]["rate"].iloc[0]
  fwd_addl_chrg=cour_rates[(cour_rates["type"]=="fwd" ) & (cour_rates["zone_code"]==Courier_zone) & (cour_rates["rate_type"]=="additional")]["rate"].iloc[0]

  amount=0
  if order_type=="Forward charges":
    
    amount+=fwd_base_chrg
    amount+=(fwd_addl_chrg*(weight_sec-1))
    return amount
  elif order_type=="Forward and RTO charges":
    rto_base_chrg=cour_rates[(cour_rates["type"]=="rto" ) & (cour_rates["zone_code"]==Courier_zone) & (cour_rates["rate_type"]=="fixed")]["rate"].iloc[0]
    rto_addl_chrg=cour_rates[(cour_rates["type"]=="rto" ) & (cour_rates["zone_code"]==Courier_zone) & (cour_rates["rate_type"]=="additional")]["rate"].iloc[0]

    amount+=fwd_base_chrg
    amount+=(fwd_addl_chrg*(weight_sec-1))
    amount+=rto_base_chrg
    amount+=(rto_addl_chrg*(weight_sec-1))
    return amount

df["Charges Billed by Courier Company (Rs.)"]=df["Order ID"].apply(get_courier_charges)
df

Order ID     AWB Number  Total weight as per X KG  \
0    2001806232  1091117222124                     1.302   
1    2001806273  1091117222194                     0.615   
2    2001806408  1091117222931                     2.265   
3    2001806458  1091117223244                     0.700   
4    2001807012  1091117229345                     0.240   
..          ...            ...                       ...   
119  2001812941  1091118551656                     0.500   
120  2001809383  1091117614452                     0.607   
121  2001820978  1091120922803                     0.515   
122  2001811475  1091121844806                     0.689   
123  2001811305  1091121846136                     0.750   

     Weight slab as per X KG  Total weight as per Courier Company KG  \
0                        1.5                                    1.30   
1                        1.0                                    1.00   
2                        2.5                                    2.50   
3                        1.0                                    1.00   
4                        0.5                                    0.15   
..                       ...                                     ...   
119                      0.5                                    0.73   
120                      1.0                                    0.50   
121                      1.0                                    0.50   
122                      1.0                                    0.50   
123                      1.0                                    0.50   

      Weight slab charged by Courier Company KG Company_Zone Courier_Zone  \
0                                           1.5            d            d   
1                                           1.0            d            d   
2                                           2.5            d            d   
3                                           1.0            b            b   
4                                           0.5            d            d   
..                                          ...          ...          ...   
119                                         1.0            b            b   
120                                         0.5            b            b   
121                                         0.5            b            b   
122                                         0.5            e            e   
123                                         0.5            b            b   

     Expected Charge as per X (Rs.)  Charges Billed by Courier Company (Rs.)  
0                             135.0                                    135.0  
1                              90.2                                     90.2  
2                             224.6                                    224.6  
3                              61.3                                     61.3  
4                              45.4                                     45.4  
..                              ...                                      ...  
119                            33.0                                     61.3  
120                           110.1                                     53.5  
121                            61.3                                     33.0  
122                           112.1                                     56.6  
123                            61.3                                     33.0  

[124 rows x 10 columns]

In [ ]:
df["Difference Between Expected Charges and Billed Charges (Rs.)"]=df["Expected Charge as per X (Rs.)"]-df["Charges Billed by Courier Company (Rs.)"]


In [ ]:
df["Difference Between Expected Charges and Billed Charges (Rs.)"].sum()

-2065.7999999999997

In [ ]:
df.to_csv("/content/drive/MyDrive/cointab_assignment/Ans_v!")